src/theory/splitter.cpp

```diff
@@ -0,0 +1,181 @@
+    : d_numPartitions(options::computePartitions()),
+      d_numChecks(0),
+      d_numPartitionsSoFar(0),
+      d_partitionFile(options::writePartitionsToFileName())
+{
+  d_valuation = std::make_unique<Valuation>(theoryEngine);
+  d_propEngine = propEngine;
+  d_output = &std::cout;
+  if (d_partitionFile != "")
+  {
+    d_partitionFileStream.open(d_partitionFile);
+    d_output = &d_partitionFileStream;
+    d_partitionFileStream.close();
+  }
+  if (options::partitionConflictSize() == 0)
+  {
+    d_conflictSize = (unsigned)log2(d_numPartitions);
+  }
+  else
+  {
+    d_conflictSize = options::partitionConflictSize();
+  }
+}
+
+void Splitter::collectDecisionLiterals(std::vector<TNode>& literals)
+{
+  std::vector<Node> decisionNodes = d_propEngine->getPropDecisions();
+  for (Node n : decisionNodes)
+  {
+    Node originalN = SkolemManager::getOriginalForm(n);
+
+    // Make sure the literal does not have a boolean term or skolem in it.
+    std::unordered_set<Kind, kind::KindHashFunction> kinds = {
+        kind::SKOLEM, kind::BOOLEAN_TERM_VARIABLE, kind::CONST_BOOLEAN};
+
+    // If the literal is the not of some node, do the checks for the child
+    // of the not instead of the not itself.
+    Node original = originalN.getKind() == kind::NOT ? originalN[0] : originalN;
+    if (expr::hasSubtermKinds(kinds, original)
+        || !d_valuation->isSatLiteral(original)
+        || !d_valuation->isDecision(original)
+        || Theory::theoryOf(original) == THEORY_BOOL)
+    {
+      continue;
+    }
+
+    literals.push_back(originalN);
+  }
+}
+
+void Splitter::emitCube(Node toEmit)
+{
+  *d_output << toEmit << "\n";
+  ++d_numPartitionsSoFar;
+}
+
+void Splitter::closeFile()
+{
+  if (d_partitionFile != "")
+  {
+    d_partitionFileStream.close();
+  }
+}
+
+TrustNode Splitter::blockPath(Node toBlock)
+{
+  // Now block the path in the search.
+  Node lemma = toBlock.notNode();
+  d_assertedLemmas.push_back(lemma);
+  TrustNode trustedLemma = TrustNode::mkTrustLemma(lemma);
+  return trustedLemma;
+}
+
+// Send lemma that is the negation of all previously asserted lemmas.
+TrustNode Splitter::stopPartitioning()
+{
+  Node lemma = NodeManager::currentNM()->mkAnd(d_assertedLemmas).notNode();
+  return TrustNode::mkTrustLemma(lemma);
+}
+
+TrustNode Splitter::makePartitions()
+{
+  d_numChecks = d_numChecks + 1;
+
+  if (d_numChecks < options::numChecks())
+  {
+    return TrustNode::null();
+  }
+
+  if (d_partitionFile != "")
+  {
+    d_partitionFileStream.open(d_partitionFile, std::ios_base::app);
+    d_output = &d_partitionFileStream;
+  }
+
+  // This is the revised version of the old splitting strategy.
+  // Cubes look like the following:
+  // C1 = l1_{1} & .... & l1_{d_conflictSize}
+  // C2 = l2_{1} & .... & l2_{d_conflictSize}
+  // C3 = l3_{1} & .... & l3_{d_conflictSize}
+  // C4 = !C1 & !C2 & !C3
+  if (options::partitionStrategy() == options::PartitionMode::REVISED)
```

Comment:
If we foresee adding other partitioning strategies as well, I'd suggest factoring this out into some helper function or class.

Reply:
There are a bunch of partitioning strategies. What sort of helper function/class did you have in mind?

Reply:
I don't really care, as long as the code is not inlined into the `makePartition()` function. Given that the strategies seem to use common helper functions, it probably makes sense to simply add private methods like `partitionRevised()` etc.